profiling/src/allocation.rs

```diff
@@ -89,122 +89,126 @@ thread_local! {
-                        locals.profiling_allocation_enabled = false;
-                    });
-                } else {
-                    info!("Memory allocation profiling enabled.")
-                }
-            }
+    if !is_zend_mm() {
+        // Neighboring custom memory handlers found
+        debug!("Found another extension using the ZendMM custom handler hook");
+        unsafe {
+            zend::zend_mm_get_custom_handlers(
+                zend::zend_mm_get_heap(),
+                &mut PREV_CUSTOM_MM_ALLOC,
+                &mut PREV_CUSTOM_MM_FREE,
+                &mut PREV_CUSTOM_MM_REALLOC,
+            );
+            ALLOCATION_PROFILING_ALLOC = allocation_profiling_prev_alloc;
+            ALLOCATION_PROFILING_FREE = allocation_profiling_prev_free;
+            ALLOCATION_PROFILING_REALLOC = allocation_profiling_prev_realloc;
+        }
+    } else {
+        unsafe {
+            ALLOCATION_PROFILING_ALLOC = allocation_profiling_orig_alloc;
+            ALLOCATION_PROFILING_FREE = allocation_profiling_orig_free;
+            ALLOCATION_PROFILING_REALLOC = allocation_profiling_orig_realloc;
         }
     }
+
+    // install our custom handler to ZendMM
+    unsafe {
+        zend::ddog_php_prof_zend_mm_set_custom_handlers(
+            zend::zend_mm_get_heap(),
+            Some(alloc_profiling_malloc),
+            Some(alloc_profiling_free),
+            Some(alloc_profiling_realloc),
+        );
+    }
+
+    // `is_zend_mm()` should be `false` now, as we installed our custom handlers
+    if is_zend_mm() {
+        error!("Memory allocation profiling could not be enabled. Please feel free to fill an issue stating the PHP version and installed modules. Most likely the reason is your PHP binary was compiled with `ZEND_MM_CUSTOM` being disabled.");
+        REQUEST_LOCALS.with(|cell| {
+            let mut locals = cell.borrow_mut();
+            locals.profiling_allocation_enabled = false;
+        });
+    } else {
+        trace!("Memory allocation profiling enabled.")
+    }
 }
 
 pub fn allocation_profiling_rshutdown() {
     REQUEST_LOCALS.with(|cell| {
         let mut locals = cell.borrow_mut();
 
-        {
-            if locals.profiling_allocation_enabled {
-                // If `is_zend_mm()` is true, the custom handlers have been reset to `None`
-                // already. This is unexpected, therefore we will not touch the ZendMM handlers
-                // anymore as resetting to prev handlers might result in segfaults and other
-                // undefined behaviour.
-                if !is_zend_mm() {
-                    let mut custom_mm_malloc: Option<zend::VmMmCustomAllocFn> = None;
-                    let mut custom_mm_free: Option<zend::VmMmCustomFreeFn> = None;
-                    let mut custom_mm_realloc: Option<zend::VmMmCustomReallocFn> = None;
+        if !locals.profiling_enabled || !locals.profiling_allocation_enabled {
+            return;
+        }
```

Comment:
I think this is okay but I suspect that we _may_ allow toggling this at runtime in the future, in which case we'd still want to undo our heap custom handlers. Since the code already detects whether the installed handlers are ours, maybe we're good to just run it anyway?